### PR TITLE
Update Model card for GPT2

### PR DESCRIPTION
**Pull request by ash-01xor**
# What does this PR do?

This PR updates the model-card for the GPT2 model, as described in https://github.com/huggingface/transformers/issues/36979, in an attempt to standardize all model-cards.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x]  Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
[documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
[here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@stevhliu

### CODE REVIEW

File: docs/source/en/model_doc/gpt2.md [hunk 1]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
-- GPT-2 was trained with a causal language modeling (CLM) objective and is therefore powerful at predicting the next
-  token in a sequence. Leveraging this feature allows GPT-2 to generate syntactically coherent text as it can be
-  observed in the *run_generation.py* example script.
-- The model can take the *past_key_values* (for PyTorch) or *past* (for TF) as input, which is the previously computed
-  key/value attention pairs. Using this (*past_key_values* or *past*) value prevents the model from re-computing
-  pre-computed values in the context of text generation. For PyTorch, see *past_key_values* argument of the
-  [`GPT2Model.forward`] method, or for TF the *past* argument of the
-  [`TFGPT2Model.call`] method for more information on its usage.
+> [!TIP]
+> Click on the GPT models in the right sidebar for more examples of how to apply GPT to different language tasks.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:49]:
```suggestion
> Click on the GPT-2 models in the right sidebar for more examples of how to apply GPT-2 to different language tasks.
```

File: docs/source/en/model_doc/gpt2.md [hunk 2]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
+
+- GPT-2 has absolute position embeddings, hence advised to pad inputs on the right rather than the left.
+- The model was trained with a causal language modeling (CLM) objective, making it excellent at predicting the next token in a sequence. This enables GPT-2 to generate coherent text, as demonstrated in the `run_generation.py` example script.
+- For efficient text generation, GPT-2 can reuse previously computed key/value attention pairs. Access this feature via the past_key_values parameter in PyTorch (see [GPT2Model.forward] method) or the past parameter in TensorFlow (see [TFGPT2Model.call] method).
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:53]: Include these in the `## Notes` section

File: docs/source/en/model_doc/gpt2.md [hunk 3]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
 - Enabling the *scale_attn_by_inverse_layer_idx* and *reorder_and_upcast_attn* flags will apply the training stability
   improvements from [Mistral](https://github.com/stanford-crfm/mistral/) (for PyTorch only).
 
-## Usage example
+The example below demonstrates how to generate text with [`Pipeline`] and [`Automodel`] class
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:57]:
```suggestion
The example below demonstrates how to generate text with [`Pipeline`] or the [`AutoModel`], and from the command line.
```

File: docs/source/en/model_doc/gpt2.md [hunk 4]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
 
-The `generate()` method can be used to generate text using GPT2 model.
+<hfoptions id="usage">
+<hfoption id="Pipeline">
 
-```python
->>> from transformers import AutoModelForCausalLM, AutoTokenizer
+```py
+from transformers import pipeline, set_seed
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:63]: No need to use a seed here.

```py
import torch
from transformers import pipeline

pipeline = pipeline(task="text-generation", model="openai-community/gpt2", torch_dtype=torch.float16, device=0)
pipeline("Hellow, I'm a language model")
```

File: docs/source/en/model_doc/gpt2.md [hunk 5]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
-...     do_sample=True,
-...     temperature=0.9,
-...     max_length=100,
-... )
->>> gen_text = tokenizer.batch_decode(gen_tokens)[0]
+generator("Hello, I'm a language model,", max_length=30, num_return_sequences=5)
 ```
+</hfoption>
+
+<hfoption id="AutoModel">
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:73]:
```suggestion
</hfoption>
<hfoption id="AutoModel">
```

File: docs/source/en/model_doc/gpt2.md [hunk 6]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
+
+```py
+from transformers import AutoModelForCausalLM, AutoTokenizer
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:76]: The code snippet for `AutoModel` should be:

```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2", torch_dtype=torch.float16, device_map="autp", attn_implementation="sdpa")
tokenizer = AutoTokenizer.from_pretrained("openai-community/gpt2")

input_ids = tokenzier("GPT2 is a model developed by OpenAI.". return_tensors="pt").to("cuda")

output = model.generate(**input_ids, cache_implementation="static")
print(tokenizer.decode(output[0], skip_special_tokens=True))
```

There should also be a `transformers-cli` example for text generation

File: docs/source/en/model_doc/gpt2.md [hunk 7]
```diff
@@ -14,92 +14,97 @@ rendered properly in your Markdown viewer.
...
+    temperature=0.9,
+    max_length=100,
+    )
+gen_text = tokenizer.batch_decode(gen_tokens)[0]
+```
+
+</hfoption>
+</hfoptions>
+
+Flash Attention 2 provides significant speedups for transformer models through optimized `CUDA` kernels for attention computation.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:97]: You can remove everything else after this and replace it with a quantization example featuring https://huggingface.co/openai-community/gpt2-xl

**Review by stevhliu — COMMENTED** [review]
Thanks! Missing a few examples like quantization and the transformers-cli, but otherwise looks to be in pretty good shape 🤗

File: docs/source/en/model_doc/gpt2.md (hunk 7 above)
Comment (ash-01xor) [reply docs/source/en/model_doc/gpt2.md:78]: Well updated the rest of the changes suggested @stevhliu. With  regards to the change suggested here,  would like to know what to remove precisely before i can add in a quantization example

**Comment by ash-01xor** [comment]
Updated as per the changes requested and added the respective examples

File: docs/source/en/model_doc/gpt2.md (hunk 7 above)
Comment (ash-01xor) [reply docs/source/en/model_doc/gpt2.md:112]: Added the quantization example

**Review by stevhliu — COMMENTED** [review]
Thanks for adding the quantization example, added a few more comments!

**Comment by ash-01xor** [comment]
Updated the file based on the recent suggestions @stevhliu

**Review by stevhliu — APPROVED** [review]
Thanks for your contribution! 🤗
